src/lib/Parser.hs

```diff
@@ -267,43 +268,57 @@ addImplicitImplicitArgs sourcePos (Just typ) ex =
 interfaceDef :: Parser [UDecl]
 interfaceDef = do
   keyWord InterfaceKW
-  -- TODO: allow multi-parameter type classes
-  interfaceName <- upperName <|> symName
-  (oneAnnBinder, pos) <- withPos annBinder
+  (tyCon, pos) <- withPos tyConDef
   keyWord WhereKW
-  record <- withSrc $ URecordTy <$> interfaceRecordFields ":"
-  let consName = mkInterfaceConsName interfaceName
-  let (varTypeName, oneAnnBinder') = case oneAnnBinder of
-        Bind v    -> (varName v, oneAnnBinder)
-        Ignore ty -> let genVar = mkName $ mkNoShadowingStr "gen" in
-            (genVar, Bind (genVar:>ty))
-  let tyCon = UConDef interfaceName $ toNest [oneAnnBinder']
-  let funDefs = mkFunDefs (pos, varTypeName, interfaceName) record
-  return $ (UData tyCon [UConDef consName (toNest [Ignore record])]) : funDefs
+  recordFieldsWithSrc <- withSrc $ interfaceRecordFields ":"
+  let (UConDef interfaceName uAnnBinderNest) = tyCon
+      record = (URecordTy . NoExt) <$> recordFieldsWithSrc
+      consName = mkInterfaceConsName interfaceName
+      ((_, varNames), _) = mapAccumR mkVarTypeName (0, []) uAnnBinderNest
+      tyCon' = UConDef interfaceName uAnnBinderNest
+      (WithSrc _ recordFields) = recordFieldsWithSrc
+      funDefs = mkFunDefs (pos, varNames, interfaceName) recordFields
+  return $ (UData tyCon' [UConDef consName (toNest [Ignore record])]) : funDefs
   where
-    mkFunDefs :: (SrcPos, Name, Name) -> UExpr -> [UDecl]
-    mkFunDefs meta ~(WithSrc _ (URecordTy (NoExt (LabeledItems items)))) =
+    mkVarTypeName
+      :: (Int, [Name]) -> UAnnBinder -> ((Int, [Name]), ())
+    mkVarTypeName (n, varNames) ann
+      | ~(Bind v) <- ann = ((n, (varName v):varNames), ())
```

Comment:
The `|` part here is unnecessary. Just replace `ann` in the line above with `~(Bind v)`.

Reply:
I'm also pretty confused as to what this function does. It looks like it's the identity on the state and simply unpacks the binders? Isn't that just something like `fmap (\(Bind v) -> v) varNames`?

Reply:
Oops, indeed. This was used to build variables that were set as `Ignore` before, but this indeed does not make any sense to use such a function when you're guaranteed to have named Binders!

Reply:
Seconded that removing this probably makes sense. Maybe something like `varNames = map (\(Bind v) -> v) $ fromList uAnnBinderNest` would be a good replacement?

Reply:
`varNames = fmap (\(Bind v) -> v) uAnnBinderNest` seemed sensible. Thanks for catching that :)